### DISCUSSION TITLE
Play only part or section of the map (practice mod)

### DISCUSSION THREAD

**Discussion by GSculerlor** [ppy/osu#13013] [Ideas]
**Describe the feature:** Sometimes players want to try to play only part of the map (where they fail or where the hard pattern appears) and not full of the map. At this point at stable, a lot of players can only create a new diff on the map that only contains a part that they want to play to do that.

**Proposal designs of the feature:** Add feature maybe like training mod or something like that, specify the start and end point/time on the map that player wants to try so players can play that specific part without creating new difficulty

**Comment by ticotaco72** [comment]
Or something like Autoplay but you can in any moment change it to normal gameplay

**Comment by c0linw** [comment]
I think it would be nice to have it implemented as a separate part of the game instead of as a mod. An ability to "scroll" through the beatmap in real time a la McOsu would be nice in a "Practice" mode alongside "Play" and "Edit" out of the main menu.

**Reply by MishterKirby** [reply]
I agree, and I feel like it should also let you loop a specific part so you can practice it until you're confident that you can play it.

**Comment by Aergwyn** [comment]
Isn't Target Practice an unimplemented mod for exactly that or am I remembering wrong?

**Reply by Vunistic** [reply]
Target practice does not do that

"The Target Practice mod removes all hit objects from a beatmap and replaces them with "targets", which are hit circles placed once every full beat. It also removes the approach circles, cuts the approach rate in half and causes the circles to grow and increase in brightness as they appear.

The goal is to hit all the targets, and missing one of them already causes the player to fail, unless No Fail (NF) is enabled. As a result, Target Practice is incompatible with Sudden Death (SD)."

https://osu.ppy.sh/wiki/en/Gameplay/Game_modifier/Target_Practice_%28lazer%29

Is completely different than practicing specific parts of a map because it changes the map completely:
"removes all hit objects from a beatmap and replaces them with "targets", which are hit circles placed once every full beat." Map is changed instead of kept the same

I could be mistaken though

**Comment by LeNitrous** [comment]
Target Practice is where random targets appear in triangles. Not the one they're looking for.

**Comment by szmarczak** [comment]
I'd like to propose fast forwarding to 10s before last failure on next try.

**Comment by developomp** [comment]
[osu!lazer 2021.1113.0](https://osu.ppy.sh/home/changelog/lazer/2021.1113.0) added a beatmap testing feature that can be used as a temporary solution for this problem.

**Reply by bdach** [reply]
not quite yet, as is it will start from the start of the map anyhow - have changes pending that will change that, though

**Comment by Feodor0090** [comment]
How settings of the mod should be implemented? Via 2 numberboxes (for start and end), or sliders?
Also, how to propertly handle their maximum values (end time must be less than beatmap length)? I tried to set `MaxValue` on bindables using `IBeatmap.BeatmapInfo.Length`, but that did not work.

**Reply by frenzibyte** [reply]
That sounds like not quite good UX, if this were to be implemented, it should include a custom frozen player that allows selecting a beginning and ending point, then a button or something to start playing the selected section. Something along the lines of that, giving the user a better way to know which part they want without capturing the current time and filling it in a textbox.

It could still be added as part of an MVP implementation.

**Reply by peppy** [reply]
It should be bar/beat based, not time (if it's going to be time based it should still snap to beats, potentially sharing components or logic with editor implementations).

**Comment by tropicaaal** [comment]
On the topic of a "Practice Mod" where one can scroll through a beatmap, it might be interesting to see some sort of keyboard control for live-updating speed, so that someone can practice a specific area at a different BPM to learn it, then gradually speed it up. Alternatively, there could be a UI similar to the difficulty adjust mod added in-game with the practice mod that lets you adjust ar/speed/od/cs/etc...

**Comment by Ekischleki** [comment]
Sadly, this request doesn't seem active anymore, even though a lot of people I asked would like to see a "Practice Mode" in Osu.
I have some definite ideas for the implementation of this mode. 
I suggest that one could get into practice mode via the main menu > play > practice, or get redirected to practice mode via a context menu.
A practice mode like GeometryDash's (where you have checkpoints, which can either get placed manually, or automatically, with an option to disable automatic placement) would be a good start. There are however some improvements that I'd suggest to this system. After failing, or wanting to restart and getting sent back to the most recent checkpoint, it should be that the player doesn't get dropped into gameplay right away, but instead has a small adjustment phase, where they would see gameplay of themselves, or auto mod for a few seconds, and are able to get back into playing, by clicking on a hit object. As in the original proposal requested, the ability to select a start and ending point for practice mode would be another good addition.
Another thing that could improve the UX, is suggesting a practice passage after repeated failures on certain points in the beatmap.


If you like this idea, I'd love to try and implement it. As I don't have much experience with the codebase of lazer, I'd need some help.

**Reply by BlockHead007** [reply]
That is something like adofai does and i want to see it too

**Reply by Ekischleki** [reply]
Before this can be implemented, we need to wait for green light from the core devs

**Reply by bdach** [reply]
@Ekischleki people [have tried to implement this as one of their first contributions to the codebase and failed](https://github.com/ppy/osu/pull/21408). this is a feature that we do want to happen, but we're going to be particular about UX and it must not be broken which means it's not going to be an easy ride to implement that. i wouldn't try that as a first contribution.

**Reply by Ekischleki** [reply]
@bdach Thanks for your insight. I see myself, that this is a great task to implement.
Still, I think practice mode would be a nice addition to lazer. I hope, that it can be in the game some day.

**Comment by crux-03** [comment]
Something I came to think of while playing is the idea of "Community Practice Difficulties". Some video/media platforms allows the community to set custom timestamps within a video to mark that portion as being skippable. It's used for some modern anime websites to automatically skip openings/endings. Similarly, I think osu! could benefit from the same idea - allow the community to set timestamps for a portion of the beatmap where the difficulty spike appears, and have it accessible within a dropdown menu inside osu! as a "practice difficulty".

Structurally, I think these community-made beatmap sections should have a title, and should implement a voting system so more relevant (and serious) practice difficulties appear at the top. With this, the community can either upvote or downvote a practice difficulty depending on how useful they find it.

**Comment by rainingfishy** [comment]
I would love to see the ability to immediately hop into a practice mode upon failing that navigates to just before you fail, with the option to then scrub forwards/backwards x amount of seconds.
A tool that map makers could be given to utilize would be to create named sections that players would be able to hop in and practice, similar to the guitar hero practice feature.
The training mode could have its own section within the game, should the mode prove to be popular, map makers could create beatmaps for specific patterns/genres/playstyles for certain rank ranges.. These already exist in the form of e.g. "Star training maps" "Jump training maps" but still suffer from the inability to pinpoint the exact point of the map the player wishes to work on. (without using editor knowledge which a lot of newer players aren't familiar with)
This could open the door for weekly/monthly/all time upvoted community training maps, players select a skillset they are wanting to work on and can choose varying difficulties of that skillset
Rocket league has a similar issue where there are many different skills and niche playstyles to learn which can benefit gameplay and their built in training mode does a great job of tackling this.

**Comment by Tsunari** [comment]
Will this be implemented anytime soon? I also think it would be a serious enhancement to the gameplay.

**Comment by monochrome22** [comment]
The practice mode would be accessed through the Practice Mode button at the footer, which was not implemented in SSv2 yet.

Here's a mockup showing what functionality the setup screen could have. It's based on the editor.

![Practice Mode mockup](https://github.com/user-attachments/assets/cfe7c9cd-d654-4a8a-96e7-059bdc403211)

Scrolling should snap to the next note, not beat divisor. Notes that are included in the practice section are highlighted in the gameplay preview and in the object timeline. Selecting notes in the timeline or gameplay preview would set those notes as the practice section, with the first note timing being set as the Start time and last note timing as the End time. Unlike the editor, it should be possible to drag the sides of the yellow box in the object timeline, only visible notes would be highlighted in the gameplay preview, clicking away wouldn't deselect, and it wouldn't be possible to exclude notes between the first and the last notes. 
When first entering the practice mode, by default, the entire map would be set as the practice section, from the first note to the last. It wouldn't be possible to not have something selected.
The ability to analyze the map would be useful for practising, that's why I included the object timeline, beat divisor display setting, and the BPM in the bottom left corner. Also, currently, players enter the editor to analyze the map. It would be more appropriate to do that in the practice tool.

The player should be able to view the replay, so that they can figure out what they're doing wrong and to keep track of progress. Ideally, there would be a "Replay history" list in the practice setup screen. Not a leaderboard, but a replay history, though it would work basically the same way. 
If the user sets a certain section with timestamps, then the replays will appear if there are any. Adjusting the section timestamps would cause the replay history list to disappear if there are no replays for that new section. 
Hovering a play in replay history would show the additional information card, same as when hovering a score in song select leaderboard, and right clicking a play would allow deleting it. 
The replay should save in history as soon as the player hits the last note, and not when they go to the results screen. Meaning, if the player quick restarts right after hitting the last note, or they are using the auto restart option, the replay would still save. 
There should be an option to immediately view the replay when pausing. Because if, for example, the player sets a 1 minute section, but they miss 20 seconds into the play, they might want to immediately view the replay to see what they did wrong. It would be bad if they had to sit there for 40 seconds waiting for the map to end. If the player views this incomplete replay, then in the unplayed section, the replay would either continue with no inputs, or the play would just fail at where they paused. It would still be saved in replay history. 

There should be an ease-in time, same as when playing a map normally. The play should not start immediately at the set point, like it does in the editor, since that makes it impossible to hit the first object due to how abrupt the start is.
There could be a setting for adjusting ease-in time, since different ease-in times can work better for different patterns and ARs.

Having presets would be useful, so that the player doesn't need to set the sections and mods every time. The player could create presets for "Ending jumps part" , "Sliders part", "Stream part", etc. 
Presets would be per beatmap. When creating a preset, a popup box would ask to input the name. 

Mods currently being used in song select should persist when entering the practice mode.

By default, the practice should not automatically restart, but there should be an option to enable it. If automatic restarting is disabled, then finishing the section would put the player at the results screen. The results screen would be the same as normal. Having a results screen is important, because it allows the player to view detailed statistics.

By default, the play should not fail, but there should be an option to allow failing, so that players can go for trying to pass difficult sections without failing. If the user enables the No Fail mod, then the "enable failing" option would untick, and ticking it would deselect the No Fail mod.

The plays made in Practice Mode shouldn't appear in regular leaderboards in any way and it might be better to disable exporting replays.

It might be good to also display the beatmap stats somewhere in the setup screen; the CS, AR, OD, HP. The BPM, length, and star rating could dynamically adjust based on the selected section.

When the player plays a map normally, then quits or fails, then enters the Practice Mode, it would be good if the timeline was already scrolled to the point where they quit or failed.

The point of this post is to suggest a direction for the UX of the Practice Mode feature, feel free to make changes as needed.

**Comment by Banana090** [comment]
Some another comment suggested implementing "map preview" in song select. Like holding "space" button hides the UI and shows you the actual map gameplay along with the song. Which I think is pretty cool idea and can be merged with this by adding "Try/Play" button on map preview screen and clickable timeline to scroll

**Comment by LFACC0** [comment]
Oh... I was wondering why the practice mode hadn't been implemented yet.
In my humble opinion as a rhythm game user, I would like to see something as direct and simple like the Geometry Dash practice mode.

in order to travel and arrive to the part of the map that you want to practice, the player has to play the map from the start and press a checkpoint button. The checkpoint marks the player's position in-game, letting the player know exactly where and when the game will restart in case of fail or fast-retry.

With this, mixed with the editor's testing mode, should be enough to form a basic "practice mode" environment, where the player can practice a map directly from the song selector.

As an addition to this feature, It would be cool to add something similar to the "bookmarks" available on the editing mode, where the player can store and save local timestamps as checkpoints. (set a maximum of timestamps if needed)

The basic behavior for this:

Loading the practice mode
1. Enter practice mode from the song selector
2.  Loading screen shows on screen
3. check if there's any checkpoint timestamps stored for the map
4. if there's one or more checkpoints, loading screen pops up a progress bar. (If the cursor doesn't hover over, load the gameplay screen as usual)
5. skip previous step if there's no timestamps saved, load the gameplay screen as usual.

<img width="1135" height="731" alt="imagen" src="https://github.com/user-attachments/assets/a2f35a7f-8d4e-4ee7-8006-859483f96819" />

Practice mode gameplay:
1. Load the gameplay screen as usual
2. Start from the beginning if no checkpoint timestamp was selected on the loading screen
3."Place a checkpoint" - pressing a hotkey places a checkpoint, (if there's a checkpoint 5 or 10 seconds before or after, replace it) (if the checkpoint timestamps limit it's exceeded, replace the nearest one)
4. A argon-like progress bar appears on the top side of the screen, showing the position of the stored checkpoint timestamps.
From here, three things can occur:
5. "Go to the next checkpoint" - Press a hotkey to stop the gameplay and jump to the expected checkpoint  thru "fast retry" animation
6. "Go to the previous checkpoint" - press a hotkey (or fast retry hotkey) to stop the gameplay and jump to the expected checkpoint  thru "fast retry" animation
7. "Go to pause Screen" - press "Esc", and the same progress bar in the gameplay screen appears. 
8. Here, right-click to delete a checkpoint, left-click to trigger fast retry and go to the clicked checkpoint. 

"Result" screen: (in case the player finished the map)
<img width="884" height="574" alt="imagen" src="https://github.com/user-attachments/assets/73fcbb3c-a595-44b5-b7c0-558eff15201b" />

That's the idea, I hope this make sense. I apologize beforehand for any confusion caused by my english writing ^^

**Comment by monochrome22** [comment]
Another simpler solution would be to just make it more convenient to create a practice difficulty. The reason why people are asking for a practice mode feature is because it's too inconvenient to create a new practice difficulty through the editor.

Right clicking the beatmap or opening beatmap Options menu would have an option for "Create practice difficulty", and it would put the player at a screen where they can select the section and create a new difficulty in song select.

<img width="204" height="322" alt="create practice difficulty" src="https://github.com/user-attachments/assets/45a3439e-559a-490e-b2cd-91fd1d50a9f8" />

<img width="1920" height="1080" alt="new diff mockup" src="https://github.com/user-attachments/assets/903f35b1-69b5-461c-930e-43857c29ab48" />

This setup screen is basically a very simplified version of the editor, I would imagine most of the code could be copy-pasted from there.

The selection would work the same way as how I described in my other post.

> Scrolling should snap to the next note, not beat divisor. Notes that are included in the practice section are highlighted in the gameplay preview and in the object timeline. Selecting notes in the timeline or gameplay preview would set those notes as the practice section, with the first note timing being set as the Start time and last note timing as the End time. Unlike the editor, it should be possible to drag the sides of the yellow box in the object timeline, only visible notes would be highlighted in the gameplay preview, clicking away wouldn't deselect, and it wouldn't be possible to exclude notes between the first and the last notes.
When first entering the practice mode, by default, the entire map would be set as the practice section, from the first note to the last. It wouldn't be possible to not have something selected.

Also

>When the player plays a map normally, then quits or fails, then enters the Practice Mode, it would be good if the timeline was already scrolled to the point where they quit or failed.

In song select, the practice difficulty would appear under the original difficulty when the original difficulty is selected. The practice difficulty should appear even if it's outside of filter range. Creating a practice difficulty would not affect beatmap status (Ranked).

<img width="872" height="619" alt="practice diff in song select" src="https://github.com/user-attachments/assets/790de2a6-474a-4fef-816d-fbb92180fb11" />

The suggestion I made above is what a proper fully functional practice mode should look like, but it would be a lot more work.

**Reply by bdach** [reply]
None of this is any simpler implementation wise.

**Comment by Vin-Renoux** [comment]
Instead of a "practice mode", I think adding a Start Position in the beatmap would be simpler implementation wise. It could be an unranked mod, or an option that appears when you right click a beatmap difficulty, in which you can set a start position based on time or beat via an input field. It should also ofcourse only produce a local score and not affect Leaderboards or ranks.

**Reply by Vin-Renoux** [reply]
Forgot to mention, it can also be like adofai's practice mode, but I imagine that would be harder to implement. I believe this feature would bring a lot of users to lazer as well if implemented.

**Reply by EthansGuild** [reply]
agreed, this would make the game unbelievably better

**Reply by Timemadman** [reply]
Start pos sounds like a great idea, for choosing where the start pos can be put via time an option to start at the chorus would also probably be great for QoL as most ppl would most likely want to start at the drop. An End pos would also be useful as ppl may not want to play the map to end from wherever they start

**Reply by Vin-Renoux** [reply]
> Start pos sounds like a great idea, for choosing where the start pos can be put via time an option to start at the chorus would also probably be great for QoL as most ppl would most likely want to start at the drop. An End pos would also be useful as ppl may not want to play the map to end from wherever they start

yea especially good for maps with diff spikes or really long maps that only have diffcult sections in the middle or further in the map.

edit: not sure about multiple start positions that may be difficult to implement but idk im not a dev. multiple start positions that you can switch between would be preferred and ideal of course

**Reply by LFACC0** [reply]
I didn't understand this idea at first. Like, how it would look this "mod" ? how is going to be the user experience, it will be comfy to interact with?

Besides that, Adofai's practice mode looks nice with the clickable diff graph in the pause screen. you just click, and then the map starts from that part and it counts down with a little fade-in of the song to just react in time

**Reply by Vin-Renoux** [reply]
Yea adofai's practice mode is cool but was mentioned/suggested before and I don't think it is as simple to implement as a Start Position. The UI/UX could just be a mod and then in the customise drop down field you can input the value in seconds for when to start the map as an example. Not complicated.

**Reply by AngelBePro** [reply]
I have made a pr just for that https://github.com/ppy/osu/pull/38345. If you really want to, you can go ahead and clone my repo and just compile it, its under the conversion category.
